File: Typescript/rate-limiter/src/algorithms/TokenBucket.ts
import { Store } from "../stores/Store";
import { RateLimitResult } from "../core/RateLimitResult";

export class TokenBucket {
    constructor(
        private store: Store,
        private capacity: number,
        private refillRatePerSec: number
    ) { }

    async allow(key: string): Promise<RateLimitResult> {
        const now = Date.now();
        const bucketKey = `bucket:${key}`;
        const lastRefillKey = `bucket:${key}:ts`;

        const tokens = (await this.store.get(bucketKey)) ?? this.capacity;
        const lastRefill = (await this.store.get(lastRefillKey)) ?? now;

        const elapsed = (now - lastRefill) / 1000;
        const newTokens = Math.min(
            this.capacity,
            tokens + elapsed * this.refillRatePerSec
        );

        if (newTokens < 1) {
            return { allowed: false, remaining: 0, retryAfterMs: 1000 };
        }

        await this.store.set(bucketKey, newTokens - 1, 60000);
        await this.store.set(lastRefillKey, now, 60000);

        return {
            allowed: true,
            remaining: Math.floor(newTokens - 1)
        };
    }
}

File: Typescript/rate-limiter/redis/token_bucket.lua
-- KEYS[1] = bucketKey
-- KEYS[2] = timestampKey
-- ARGV[1] = capacity
-- ARGV[2] = refillRate
-- ARGV[3] = now (ms)

local tokens = tonumber(redis.call("GET", KEYS[1])) or tonumber(ARGV[1])
local lastTs = tonumber(redis.call("GET", KEYS[2])) or tonumber(ARGV[3])

local elapsed = (ARGV[3] - lastTs) / 1000
local newTokens = math.min(ARGV[1], tokens + elapsed * ARGV[2])

if newTokens < 1 then
  return {0, math.floor(newTokens)}
end

redis.call("SET", KEYS[1], newTokens - 1, "PX", 60000)
redis.call("SET", KEYS[2], ARGV[3], "PX", 60000)

return {1, math.floor(newTokens - 1)}
